dll/win32/browseui/CAutoComplete.cpp

```diff
@@ -25,664 +25,2124 @@
+{
+    TRACE("CACEditCtrl::OnGetDlgCode(%p)\n", this);
+    ATLASSERT(m_pDropDown);
+
+    LRESULT ret = DefWindowProcW(uMsg, wParam, lParam); // get default
+
+    if (m_pDropDown)
+    {
+        // some special keys need default processing. we handle them here
+        switch (wParam)
         {
-            TRACE("HeapAlloc Failed when trying to alloca %d bytes\n", (wcslen(pwszQuickComplete) + 1) * sizeof(WCHAR));
-            return S_FALSE;
+            case VK_RETURN: // [Enter] key
+                if (m_pDropDown->IsWindowVisible() || ::GetKeyState(VK_CONTROL) < 0)
+                    m_pDropDown->OnEditKeyDown(VK_RETURN, 0);
+                break;
+            case VK_TAB: // [Tab] key
+                if (m_pDropDown->IsWindowVisible() && m_pDropDown->UseTab())
+                    m_pDropDown->OnEditKeyDown(VK_TAB, 0);
+                break;
+            case VK_ESCAPE: // [Esc] key
+                if (m_pDropDown->IsWindowVisible())
+                    ret |= DLGC_WANTALLKEYS; // we want all keys to manipulate the list
+                break;
+            default:
+            {
+                ret |= DLGC_WANTALLKEYS; // we want all keys to manipulate the list
+                break;
+            }
         }
     }
 
-    return S_OK;
+    return ret;
 }
 
-/**************************************************************************
- *  IAutoComplete_fnGetOptions
- */
-HRESULT WINAPI CAutoComplete::GetOptions(DWORD *pdwFlag)
+// WM_KEYDOWN
+// This message is posted to the window with the keyboard focus when a non-system key is pressed.
+LRESULT CACEditCtrl::OnKeyDown(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
 {
-    HRESULT hr = S_OK;
+    TRACE("CACEditCtrl::OnKeyDown(%p, %p)\n", this, wParam);
+    ATLASSERT(m_pDropDown);
+    if (m_pDropDown->OnEditKeyDown(wParam, lParam))
+        return 1; // eat
+    bHandled = FALSE; // do default
+    return 0;
+}
 
-    TRACE("(%p) -> (%p)\n", this, pdwFlag);
+// WM_KILLFOCUS @implemented
+// This message is sent to a window immediately before it loses the keyboard focus.
+LRESULT CACEditCtrl::OnKillFocus(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACEditCtrl::OnKillFocus(%p)\n", this);
+    ATLASSERT(m_pDropDown);
 
-    *pdwFlag = m_options;
+    // hide the list if lost focus
+    HWND hwndGotFocus = (HWND)wParam;
+    if (hwndGotFocus != m_hWnd && hwndGotFocus != m_pDropDown->m_hWnd)
+    {
+        m_pDropDown->HideDropDown();
+    }
 
-    return hr;
+    bHandled = FALSE; // do default
+    return 0;
 }
 
-/**************************************************************************
- *  IAutoComplete_fnSetOptions
- */
-HRESULT WINAPI CAutoComplete::SetOptions(DWORD dwFlag)
+// WM_PASTE @implemented
+// An application sends this message to an edit control to paste at the current caret position.
+LRESULT CACEditCtrl::OnPaste(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
```

Comment:
Again this function looks identical to the ones for WM_CUT and WM_CLEAR.

Reply:
United.